File: 2360. Longest Cycle in a Graph/Solution.java
class Solution {
    int[] visited;
    public int longestCycle(int[] edges) {
        int n = edges.length, res = -1;
        visited = new int[n];
        Arrays.fill(visited, -1);
        for (int i = 0; i < n; i++) {
            if (visited[i] == -1) res = Math.max(res, dfs(edges, i, i, 0));
        }
        return res;
    }

    public int dfs(int[] edges, int curr, int start, int depth) {
        if (curr == -1) return -1;
        if (start == curr && depth > 0) return depth;
        if (visited[curr] != -1) return -1;
        visited[curr] = 0;
        visited[curr] = dfs(edges, edges[curr], start, depth + 1);
        return visited[curr];
    }
}

File: 2360. Longest Cycle in a Graph/Solution.py
class Solution:
    def longestCycle(self, edges: List[int]) -> int:
        n, res = len(edges), -1
        self.visited = [False for i in range(n)]
        for i in range(n):
            if not self.visited[i]: res = max(res, self.dfs(edges, i, i, 0))
        return res
    
    def dfs(self, edges, curr, start, depth):
        if curr == -1: return -1
        if curr == start and depth > 0: return depth
        if self.visited[curr]: return -1
        self.visited[curr] = True
        res = self.dfs(edges, edges[curr], start, depth + 1)
        if res == -1: self.visited[curr] = False
        return res